airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py

```diff
@@ -0,0 +1,527 @@
+            stream_name=child_name,
+            output_directory=parent.output_directory,
+            integration_type=parent.integration_type,
+            raw_schema=parent.raw_schema,
+            schema=parent.schema,
+            json_column_name=json_column_name,
+            properties=properties,
+            tables_registry=parent.tables_registry,
+        )
+        self.parent = parent
+        self.is_nested_array = is_nested_array
+        self.json_path = parent.json_path + [child_name]
+        return self
+
+    def init(
+        self,
+        stream_name: str,
+        output_directory: str,
+        integration_type: DestinationType,
+        raw_schema: str,
+        schema: str,
+        json_column_name: str,
+        properties: Dict,
+        tables_registry: Set[str],
+    ):
+        """
+        @param stream_name of the stream being processed
+
+        @param output_directory is the path to the directory where this processor should write the resulting SQL files (DBT models)
+        @param integration_type is the destination type of warehouse
+        @param raw_schema is the name of the staging intermediate schema where to create internal tables/views
+        @param schema is the name of the schema where to store the final tables where to store the transformed data
+
+        @param json_column_name is the name of the column in the raw data table containing the json column to transform
+        @param properties is the json schema description of this stream
+
+        @param tables_registry is the global context recording all tables created so far
+        """
+        self.stream_name = stream_name
+        self.output_directory = output_directory
+        self.integration_type = integration_type
+        self.raw_schema = raw_schema
+        self.schema = schema
+        self.json_column_name = json_column_name
+        self.properties = properties
+        self.tables_registry = tables_registry
+
+        self.name_transformer = DestinationNameTransformer(integration_type)
+        self.json_path = [stream_name]
+        return self
+
+    def process(self, from_table: str) -> Dict:
+        """
+        @param from_table refers to the raw source table to use to extract data from
+        """
+        # Check properties
+        if not self.properties:
+            print(f"  Ignoring substream '{self.stream_name}' from {self.current_json_path()} because properties list is empty")
+            return
+
+        # Transformation Pipeline for this stream
+        from_table = self.write_model(self.generate_json_parsing_model(from_table), is_intermediate=True, suffix="ab1")
+        from_table = self.write_model(self.generate_column_typing_model(from_table), is_intermediate=True, suffix="ab2")
+        from_table = self.write_model(self.generate_id_hashing_model(from_table), is_intermediate=True, suffix="ab3")
+        from_table = self.write_model(self.generate_final_model(from_table), is_intermediate=False)
```

Comment:
the idea here is each of these models queries the previous model?

Reply:
yes, they are chained together from the raw table in green to the final table in blue:

![Screenshot 2021-02-19 at 11 47 36](https://user-images.githubusercontent.com/7718175/108494834-499c8a00-72a8-11eb-9e73-e649b0a7ac73.png)